src/data/set/function.lean

```diff
@@ -418,6 +432,17 @@ lemma surj_on.maps_to_compl (h : surj_on f s t) (h' : injective f) : maps_to f s
 lemma maps_to.surj_on_compl (h : maps_to f s t) (h' : surjective f) : surj_on f sᶜ tᶜ :=
 h'.forall.2 $ λ x ht, mem_image_of_mem _ $ λ hs, ht (h hs)
 
+lemma eq_on.cancel_right (hf : s.eq_on (g₁ ∘ f) (g₂ ∘ f)) (hf' : s.surj_on f t) : t.eq_on g₁ g₂ :=
+begin
+  intros b hb,
+  obtain ⟨a, ha, rfl⟩ := hf' hb,
+  exact hf ha,
+end
+
+lemma surj_on.cancel_right (hf : s.surj_on f t) (hf' : s.maps_to f t) :
```

Comment:
```suggestion
lemma surj_on.cancel_right (hf : f '' s = t) :
```
Don't we prefer this formulation? Also, I'm missing the `iff` stating that these two are equivalent.

Reply:
This doesn't work as nicely with dot notation, so I'm adding this lemma instead:
```
lemma eq_on_comp_right_iff : s.eq_on (g₁ ∘ f) (g₂ ∘ f) ↔ (f '' s).eq_on g₁ g₂ :=
(s.surj_on_image f).cancel_right $ s.maps_to_image f
```